File: backend/app/schemas/share.py
"""Wire shapes for the public share surface (roadmap 1.1).

The record payload embeds the existing timeline keys (``events``,
``biomarkers``, ``visits``, ``instrumental``) unchanged, so the frontend's
timeline types and components apply to a shared record with no second
vocabulary. ``meta`` and ``header`` are the only additions; they are what the
recipient's orientation strip renders.
"""

from typing import Optional

from pydantic import BaseModel, Field

from app.schemas.biomarker import BiomarkerDefinition, BiomarkerResult
from app.schemas.common import TimelineResponse


class SharedBiomarkerDefinition(BiomarkerDefinition):
    """Owner-facing verification flag: the recipient has no use for it, so it
    is dropped from the public payload rather than sent as False (technical
    plan §5.2/§10)."""

    canonical_unit_inferred: bool = Field(default=False, exclude=True)


class SharedBiomarkerResult(BiomarkerResult):
    definition: SharedBiomarkerDefinition


class ShareRecordMeta(BaseModel):
    """The orientation strip's data: how fresh the record is and how long the
    link lives."""

    created_at: str
    expires_at: str
    last_updated: str
    # {"kind": "all"} or {"kind": "range", "from": ..., "to": ...}
    scope: dict
    default_locale: Optional[str] = None


class ShareRecordHeader(BaseModel):
    """The owner's personal header, present only when the link includes it and
    the owner is a registered principal (an anonymous session has no name or
    date of birth to share)."""

    name: str = ""
    dob: str = ""
    gender: str = ""


class SharedRecordResponse(TimelineResponse):
    meta: ShareRecordMeta
    header: Optional[ShareRecordHeader] = None
    biomarkers: list[SharedBiomarkerResult]


class ShareLinkCreatedResponse(BaseModel):
    """The create response. ``token`` is the raw value and is returned exactly
    once — only its hash is stored.

    ``include_notes`` is deliberately absent (Stage 2, S3): entry free-text
    notes never travel on a shared record, so the toggle left the API surface
    and the DB column behind it is unread.
    """

    id: str
    token: str
    created_at: str
    expires_at: str
    scope: dict
    include_header: bool
    default_locale: Optional[str] = None
    # Whether the link needs a passcode — never the passcode, never its hash.
    # The recipient's own page is the only place the code is ever typed.
    requires_passcode: bool = False


class ShareUnlockRequest(BaseModel):
    """The unlock body: the raw link token and the code the recipient typed.

    Both travel in a POST body rather than the URL, for the same reason the
    read token travels in a header — this repo has no log-scrubbing layer, and
    whatever is in a path is written to stdout."""

    # Both default to "" so a body that OMITS either key (or sends null) takes
    # the SAME uniform 400 the wrong-code path takes, instead of a 422 that
    # echoes the token back in a validation detail. An empty code can never
    # match a stored hash, so this cannot let anything in (Stage 4 review, F4).
    # A body that is not valid JSON is still refused by request validation —
    # the plan's §4 sentence was corrected to say exactly that.
    token: Optional[str] = ""
    passcode: Optional[str] = ""


class ShareUnlockResponse(BaseModel):
    """A short-lived unlock grant.

    A bearer credential: it expires, it dies with the link, and the recipient
    keeps it in `sessionStorage` rather than a cookie (the share surface sets
    no cookie by contract). The expiry is echoed so the client can drop the
    grant instead of waiting for a 404.
    """

    grant: str
    expires_at: str


class ShareLinkCreateRequest(BaseModel):
    """The create body (Stage 2, S4-S6): how long the link lives, what range of
    the record it may show, and whether the owner's header travels with it.

    ``expiry_days`` is validated against the principal in the router (the
    allowed sets differ for anonymous and registered senders); ``scope`` is a
    range or nothing; ``default_locale`` (Stage 3, S10) is the sender's
    per-link language preset, ``"en"`` / ``"ru"`` or None (the recipient's
    browser decides). All are server-side rules, never just hidden UI.
    """

    # Whole-record default: None means {"kind": "all"}.
    expiry_days: int = 7
    scope: Optional[dict] = None
    include_header: bool = True
    default_locale: Optional[str] = None
    # Optional passcode (Stage 4, S15). The minimum length is validated in the
    # router beside the other principal-dependent rules, and the raw value is
    # hashed before it reaches the row: never stored, echoed or logged. A link
    # without one behaves exactly as it did before Stage 4.
    passcode: Optional[str] = None


class ShareLinkSummary(BaseModel):
    """One row of the sender's "Shared links" list.

    Everything the card renders is computed server-side — the client never
    derives the state, and never learns a token (only its hash exists).
    """

    id: str
    created_at: str
    expires_at: str
    revoked_at: Optional[str] = None
    first_opened_at: Optional[str] = None
    open_count: int = 0
    last_opened_at: Optional[str] = None
    is_anonymous: bool
    # {"kind": "all", "from": None, "to": None} or
    # {"kind": "range", "from": "YYYY-MM-DD"|None, "to": ...}.
    scope: dict
    include_header: bool
    # The sender's per-link language preset (Stage 3, S10): "en" / "ru" or
    # None when the recipient's browser decides.
    default_locale: Optional[str] = None
    # True when the link needs a passcode (Stage 4, S15). The card renders a
    # "protected" marker from this; the passcode itself is unrecoverable.
    requires_passcode: bool = False
    # "active" | "expired" | "revoked" — revoked wins over expired.
    state: str
    # True when the owner's record is newer than what this link has been
    # acknowledged for (never true for a revoked link).
    has_new_data: bool = False


class ShareLinkListResponse(BaseModel):
    links: list[ShareLinkSummary] = []


class ShareNoticeResponse(BaseModel):
    """The sender's new-data notice. A pure read: ``show`` is computed on
    every request and is never acknowledged as a side effect of asking."""

    active_links: int = 0
    show: bool = False


class ShareNoticeAckResponse(BaseModel):
    success: bool = True
    acknowledged: int = 0


class ShareLinkRevokedResponse(BaseModel):
    success: bool = True
    id: str


class ShareLinksRevokedResponse(BaseModel):
    success: bool = True
    revoked: int = 0


__all__ = [
    "ShareLinkCreateRequest",
    "ShareLinkCreatedResponse",
    "ShareLinkListResponse",
    "ShareLinkRevokedResponse",
    "ShareLinkSummary",
    "ShareLinksRevokedResponse",
    "ShareNoticeAckResponse",
    "ShareNoticeResponse",
    "ShareRecordHeader",
    "ShareRecordMeta",
    "ShareUnlockRequest",
    "ShareUnlockResponse",
    "SharedBiomarkerDefinition",
    "SharedBiomarkerResult",
    "SharedRecordResponse",
]

File: frontend/src/lib/share.ts
import type {
  BiomarkerDefinition,
  BiomarkerResult,
  DateHeader,
  InstrumentalData,
  MatrixCategory,
  MedicalEvent,
  VisitData,
} from './types'
import { isOutOfRange } from './status-labels'

/**
 * Types and pure helpers for the public share surface (roadmap 1.1). The
 * payload mirrors the timeline response so the real components apply to it
 * unchanged; only `meta` and `header` are new.
 */

export const SHARE_TOKEN_HEADER = 'X-Share-Token'

/**
 * The unlock grant (Stage 4, S15). A second credential for the same request:
 * the token says WHICH link, the grant says "this reader already typed the
 * code". Held in `sessionStorage`, never a cookie — the share surface sets no
 * cookie by contract.
 */
export const SHARE_GRANT_HEADER = 'X-Share-Grant'

/** Where the grant is kept: per tab, forgotten when the tab closes. */
export const SHARE_GRANT_STORAGE_KEY = 'hp.share.grant'

export type ShareEntryType = 'blood_test' | 'doctor_visit' | 'instrumental_test' | 'procedure'

/**
 * The four types a sender may exclude, in canonical order. The dialog renders
 * one checkbox each and the card reads the stored list back in words; the
 * backend refuses anything else, so this list is the vocabulary, not a
 * suggestion.
 */
export const SHARE_ENTRY_TYPES: readonly ShareEntryType[] = [
  'blood_test',
  'doctor_visit',
  'instrumental_test',
  'procedure',
] as const

/** The minimum passcode length the server enforces (S15). */
export const SHARE_PASSCODE_MIN_LENGTH = 6

/** What the shared page should render for a link, before any record is read. */
export type SharedFirstPaint = 'record' | 'protected' | 'unavailable'

/**
 * The shared page's first-paint decision (Stage 4, S15), as a pure function.
 *
 * `GET /api/share/status` answers one of three things and the page has to turn
 * it into one of three renders. Extracted from the page because a server
 * component is impractical to render in vitest, and this is the branch a real
 * browser had to catch during Stage 4 development — it should not be the one
 * piece with no test (Stage 4 review).
 */
export function sharedFirstPaint(
  status: { requires_passcode: boolean } | null,
): SharedFirstPaint {
  if (status === null) return 'unavailable'
  return status.requires_passcode ? 'protected' : 'record'
}

export interface ShareRecordMeta {
  created_at: string
  expires_at: string
  last_updated: string
  scope: ShareLinkScope
  default_locale: string | null
}

export interface ShareRecordHeader {
  name: string
  dob: string
  gender: string
}

export interface SharedRecord {
  meta: ShareRecordMeta
  header: ShareRecordHeader | null
  events: MedicalEvent[]
  biomarkers: BiomarkerResult[]
  visits: Record<string, VisitData>
  instrumental: Record<string, InstrumentalData>
}

export interface SharedFlowsheet {
  dates: DateHeader[]
  matrix: MatrixCategory[]
  biomarkers: BiomarkerResult[]
}

export interface ShareLinkSummary {
  id: string
  created_at: string
  expires_at: string
  revoked_at: string | null
  first_opened_at: string | null
  open_count: number
  last_opened_at: string | null
  is_anonymous: boolean
  scope: ShareLinkScope
  include_header: boolean
  /** The sender's per-link language preset (S10): "en" / "ru" or null. */
  default_locale: string | null
  /** True when the link needs a passcode (S15). The code itself is gone. */
  requires_passcode: boolean
  /** Computed SERVER-side — render it, never re-derive it from the dates. */
  state: ShareLinkState
  /** Server-computed: the record gained entries after the sender acknowledged. */
  has_new_data: boolean
}

export interface ShareLinkCreated {
  id: string
  token: string
  created_at: string
  expires_at: string
  scope: ShareLinkScope
  include_header: boolean
  default_locale: string | null
  requires_passcode: boolean
}

/**
 * The window a link exposes. `all` is the whole record; a `range` narrows it
 * to whole days, and either end may be omitted (open-ended).
 */
export interface ShareLinkScope {
  kind: 'all' | 'range'
  from?: string | null
  to?: string | null
  /** Entry types this link withholds (S16). Absent when nothing is excluded. */
  exclude?: ShareEntryType[] | null
}

/** The scope posted when creating a link: `null` means the whole record. */
export interface ShareScopeRangeInput {
  kind: 'range'
  from: string | null
  to: string | null
  exclude?: ShareEntryType[]
}

/**
 * `POST /api/share/links` body. `expiry_days` is validated against the
 * principal server-side (anonymous senders may only ask for 1 or 7), which is
 * why the dialog hides 30 days for them instead of letting the request 400.
 */
export interface ShareLinkCreateInput {
  expiry_days: number
  scope: ShareScopeRangeInput | ShareScopeAllInput | null
  include_header: boolean
  /** "en" / "ru", or null so the recipient's browser decides (S10). */
  default_locale: string | null
  /** Optional passcode (S15); null/omitted means the link is open. */
  passcode?: string | null
}

/** The whole record, optionally minus some entry types (Stage 4, S16). */
export interface ShareScopeAllInput {
  kind: 'all'
  exclude?: ShareEntryType[]
}

/** `GET /api/share/notice` — a pure read, never the acknowledgement. */
export interface ShareNotice {
  active_links: number
  show: boolean
}

/** `POST /api/share/notice/ack` result. */
export interface ShareNoticeAck {
  success: boolean
  acknowledged: number
}

export type ShareLinkState = 'active' | 'expired' | 'revoked'

/**
 * The "Needs attention" set: out-of-range AND reliable. A reading the app
 * could not standardise (`needs_review`) is never presented as a confident
 * flag — it still appears in the full table (D13).
 */
export function flaggedBiomarkers(biomarkers: BiomarkerResult[]): BiomarkerResult[] {
  return biomarkers
    .filter((b) => isOutOfRange(b.status) && !b.needs_review)
    .sort((a, b) => new Date(b.date).getTime() - new Date(a.date).getTime())
}

/**
 * Display name in the recipient's language, falling back to English and then
 * to the raw id. Reads the persisted multilingual name — the shared view must
 * never trigger a translation run.
 */
export function biomarkerName(definition: BiomarkerDefinition, locale: string): string {
  const names = definition.names ?? {}
  return names[locale] || names.en || definition.id
}

/** The link a recipient can open. The token travels as a path segment. */
export function shareUrl(origin: string, token: string): string {
  return `${origin.replace(/\/$/, '')}/s/${token}`
}
